src/Build/BackEnd/Shared/BuildRequestConfiguration.cs

```diff
@@ -455,6 +440,48 @@ public ProjectInstance Project
             }
         }
 
+        private void SetProjectBasedState(ProjectInstance project)
+        {
+            ErrorUtilities.VerifyThrow(project != null, "Cannot set null project.");
+            _project = project;
+            _baseLookup = null;
+
+            // Clear these out so the other accessors don't complain.  We don't want to generally enable resetting these fields.
+            _projectDefaultTargets = null;
+            _projectInitialTargets = null;
+            ProjectDefaultTargets = _project.DefaultTargets;
+            ProjectInitialTargets = _project.InitialTargets;
+
+            if (IsCached)
+            {
+                ClearCacheFile();
+                IsCached = false;
+            }
+        }
+
+        public void InitializeProject(BuildParameters buildParameters, Func<ProjectInstance> loadProjectFromFile)
+        {
+            ErrorUtilities.VerifyThrow(
+                Traits.Instance.EscapeHatches.SerializeEntireProjectInstance,
+                $"This method assumes the {nameof(ProjectInstance)} is entirely serialized");
+
+            if (_project == null || // building from file. Load project from file
+                _transferredProperties != null // need to overwrite particular properties, so load project from file and overwrite properties
+            )
+            {
+                Project = loadProjectFromFile.Invoke();
+            }
+            else
+            {
+                // projectInstance was serialized over. Finish initialization with node specific state
+
+                _project.ProjectRootElementCache = buildParameters.ProjectRootElementCache;
```

Comment:
Opinion needed.

I had to create these two ProjectInstance properties. Initially they only got into a ProjectInstance via the constructor and represent per node specific object. Now that the PI is travelling between nodes, they have to be able to be set after the PI is constructor.

Should I create a special `LateInitialize` method on ProjectInstance to give a clear meaning to this concept? Or just leave it like is, via the new properties?